disnake/enums.py

```diff
@@ -72,326 +85,1370 @@
+            return
 
+        if not isinstance(value, self.base):
+            raise TypeError(
+                f"Member {name} must be of type {self.base.__name__}, got {type(value).__name__}"
+            )
 
-def _create_value_cls(name: str, comparable: bool) -> Type[_EnumValueBase]:
-    parent = _EnumValueComparable if comparable else _EnumValueBase
-    return type(f"{parent.__name__}_{name}", (parent,), {"_cls_name": name})  # type: ignore
+        if name in self.member_map:
+            raise TypeError(f"Cannot create multiple members with the same name: {name!r}")
+        if value in self.value_map:
+            # We'll have to settle for slower value lookup in case of a dupe
+            self.member_map[name] = value
+            return
 
-
-def _is_descriptor(obj):
-    return hasattr(obj, "__get__") or hasattr(obj, "__set__") or hasattr(obj, "__delete__")
+        self.member_map[name] = value
+        self.value_map[value] = name
 
 
 class EnumMeta(type):
-    if TYPE_CHECKING:
-        __name__: ClassVar[str]
-        _enum_member_names_: ClassVar[List[str]]
-        _enum_member_map_: ClassVar[Dict[str, Any]]
-        _enum_value_map_: ClassVar[Dict[Any, Any]]
-
-    def __new__(cls, name, bases, attrs, *, comparable: bool = False):
-        value_mapping = {}
-        member_mapping = {}
-        member_names = []
-
-        value_cls = _create_value_cls(name, comparable)
-        for key, value in list(attrs.items()):
-            is_descriptor = _is_descriptor(value)
-            if key[0] == "_" and not is_descriptor:
-                continue
-
-            # Special case classmethod to just pass through
-            if isinstance(value, classmethod):
-                continue
-
-            if is_descriptor:
-                setattr(value_cls, key, value)
-                del attrs[key]
-                continue
-
-            try:
-                new_value = value_mapping[value]
-            except KeyError:
-                new_value = value_cls(name=key, value=value)
-                value_mapping[value] = new_value
-                member_names.append(key)
-
-            member_mapping[key] = new_value
-            attrs[key] = new_value
-
-        attrs["_enum_value_map_"] = value_mapping
-        attrs["_enum_member_map_"] = member_mapping
-        attrs["_enum_member_names_"] = member_names
-        attrs["_enum_value_cls_"] = value_cls
-        actual_cls = super().__new__(cls, name, bases, attrs)
-        value_cls._actual_enum_cls_ = actual_cls  # type: ignore
-        return actual_cls
-
-    def __iter__(cls):
-        return (cls._enum_member_map_[name] for name in cls._enum_member_names_)
-
-    def __reversed__(cls):
-        return (cls._enum_member_map_[name] for name in reversed(cls._enum_member_names_))
-
-    def __len__(cls):
-        return len(cls._enum_member_names_)
-
-    def __repr__(cls):
+    __is_enum_instantiated: ClassVar[bool] = False
+
+    _name_map_: ClassVar[Mapping[str, Enum]]
```

Comment:
Is there a reason to call this `_name_map_` instead of the stdlib's `_member_map_`? Not saying this should be changed, just wondering :p

Reply:
That reason would be me being dumb, I suppose. I'd been keeping other names as close to vanilla as possible so I'm not sure why I failed to do so here.

Reply:
Fixed in 662e0bdf1f0c47a3142d4214b1cbda16fb6e5b17.